app/services/document_filter.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+class DocumentFilter
+  SORT_KEYS = %w[updated_at].freeze
+  DEFAULT_SORT = "-updated_at"
+
+  attr_reader :filters, :sort, :page, :per_page
+
+  def initialize(params)
+    @filters = params[:filters].to_h.symbolize_keys
+    @sort = allowed_sort?(params[:sort]) ? params[:sort] : DEFAULT_SORT
+    @page = params[:page]
+    @per_page = params[:per_page]
+  end
+
+  def documents
+    scope = filtered_scope(Document)
+    scope = ordered_scope(scope)
+    scope.page(page).per(per_page)
+  end
+
+  def filter_params
+    filters.select { |_, value| value.present? }.tap do |params|
+      params[:sort] = sort if sort != DEFAULT_SORT
+      params[:page] = page if page != "1"
+    end
+  end
+
+private
+
+  def allowed_sort?(sort)
+    SORT_KEYS.flat_map { |item| [item, "-#{item}"] }.include?(sort)
+  end
+
+  def filtered_scope(scope)
+    filters.inject(scope) do |memo, (field, value)|
+      next memo if value.blank?
+
+      case field
+      when :title_or_url
+        memo.where("title ILIKE ? OR base_path ILIKE ?", "%#{value}%", "%#{value}%")
```

Comment:
Does `"%#{value}%"` introduce an SQLi vulnerability? I think you have to use [sanitize_sql_like](https://api.rubyonrails.org/classes/ActiveRecord/Sanitization/ClassMethods.html#method-i-sanitize_sql_like).

Reply:
Good point, I don't think it introduces a vulnerability as it uses a prepared statement, but using sanitise_sql_like will mean we don't get unexpected searches if a user makes uses of a % character.

Reply:
Good point - I've done this and added a test for it. I thought the '?' technique automatically sanitised the strings, but obviously not for %!